javascript/organizing_your_javascript_code/objects_and_object_constructors.md

```diff
@@ -59,45 +61,271 @@ const playerTwoMarker = "O";
+
+But, there is no rule saying that you can't add those functions to your object as well! A common convention is to prefix methods and properties that you don't intend other people to use with an underscore (`_`). This convention conveys to others that "These things are meant to be used internally by this object, please interact with the other available methods and properties on this object's interface instead".
+
+Let's see what that looks like!
+
+```js
+const rps = {
+  _options: ['rock', 'paper', 'scissors'],
+  _getRandomChoice() {
+    const randomIndex = Math.floor(Math.random() * 3)
+    return this._options[randomIndex]
+  },
+  _isTie(playerChoice, computerChoice) {return playerChoice === computerChoice},
+  _isPlayerWinner(playerChoice, computerChoice) {
+    if(
+      (playerChoice === 'rock' && computerChoice === 'scissors') ||
+      (playerChoice === 'paper' && computerChoice === 'rock') ||
+      (playerChoice === 'scissors' && computerChoice === 'paper')
+    ) {
+      return true
+    } else {
+      return false
+    }
+  },
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    // if an invalid choice is chosen, throw an error
+    if(!this._options.includes(playerChoice)) {
+      throw new Error(`Expected 'rock', 'paper', or 'scissors', but got ${playerChoice}`)
+    }
+    
+    // get the computer's choice
+    const computerChoice = this._getRandomChoice()
+
+    // determine the winner, apply points if necessary, and return who won
+    if(this._isTie(playerChoice, computerChoice)) {
+      return "tie"
+    } else if(this._isPlayerWinner(playerChoice, computerChoice)) {
+      this.playerScore++
+      return "player"
+    } else {
+      this.computerScore++
+      return 'computer'
+    }
+  },
+  getWinningPlayer() {
+    if(this.playerScore === this.computerScore) {
+      return 'tie'
+    } else if (this.playerScore > this.computerScore) {
+      return 'player'
+    } else {
+      return 'computer'
+    }
+  },
+  reset() {
+    this.playerScore = 0;
+    this.computerScore = 0;
+  }
 }
 ```
 
-Or, what if we aren't making a 2 player game, but something more complicated such as an online shopping site with a large inventory? In that case, using objects to keep track of an item's name, price, description and other things is the only way to go. Unfortunately, in that type of situation, manually typing out the contents of our objects is not feasible either. We need a cleaner way to create our objects, which brings us to...
+Another name for these might also be **private properties**/**private methods**, and even though object literal syntax doesn't provide a way to truly make them private, you will later learn about other methods of creating objects that *can*.
```

Comment:
I kind of wonder how useful it is to go over this kind property/method privacy. Mostly just because this more of a "legacy" approach and in modern JavaScript, you'd accomplish this through `class` or factory functions.

I know you reference "you'll learn a better way later", but I think showing the learner this has the potential to raise a lot of questions.

Reply:
I wanted this lesson to introduce the idea of using objects to organize code in the simplest way possible, and to implant some idea of how one might craft an interface for an object.

Learners are likely to be exposed to this idea of using `_` to denote private properties as they read through content the curriculum offers anyway, and I feel there is value in covering the idea up-front.

Perhaps with the RPS object, an example of how private properties are helpful could be provided, where the `playerScore`/`computerScore` could be private properties, and a function is used to retrieve them instead. This could illustrate how *if those properties were really private*, the intended state of the object could be protected.

In my experience helping learners, many of them just don't grasp the usefulness of using objects for grouping functionality *at all*. My hope would be, by having an early preface to some of the basic ideas, it'll be in the back of learners minds as they continue through this section. 

I also feel covering the idea up-front may help learners understand the tradeoffs of using object literals/constructors/factories/classes as we cover each of them.